drivers/ethernet/eth_ivshmem.c

```diff
@@ -0,0 +1,594 @@
+{
+	const struct device *dev = arg;
+	struct eth_ivshmem_dev_data *dev_data = dev->data;
+
+	k_sem_give(&dev_data->int_sem);
+}
+
+static uint64_t pcie_conf_read_u64(pcie_bdf_t bdf, unsigned int reg)
+{
+	uint64_t lo = pcie_conf_read(bdf, reg);
+	uint64_t hi = pcie_conf_read(bdf, reg + sizeof(uint32_t)/4);
+
+	return hi << 32 | lo;
+}
+
+static bool eth_ivshmem_pci_lookup_cb(pcie_bdf_t bdf, pcie_id_t id, void *cb_data)
+{
+	struct pcie_dev *pci_dev = cb_data;
+
+	if (id != PCI_ID_SIEMENS_IVSHMEM) {
+		return true;
+	}
+
+	uint32_t class = pcie_conf_read(bdf, PCIE_CONF_CLASSREV);
+
+	if (PCIE_CONF_CLASSREV_PROGIF(class) != IVSHMEM_PROTO_NET) {
+		return true;
+	}
+
+	pci_dev->bdf = bdf;
+
+	return false;
+}
+
+int eth_ivshmem_initialize(const struct device *dev)
+{
+	struct eth_ivshmem_dev_data *dev_data = dev->data;
+	const struct eth_ivshmem_cfg_data *cfg_data = dev->config;
+	int res;
+
+	struct pcie_dev pcie = {
+		.bdf = PCIE_BDF_NONE,
+		.id = PCI_ID_SIEMENS_IVSHMEM
+	};
+	struct pcie_scan_opt pcie_scan_opt = {
+		.cb = eth_ivshmem_pci_lookup_cb,
+		.cb_data = &pcie,
+		.flags = (PCIE_SCAN_RECURSIVE | PCIE_SCAN_CB_ALL),
+	};
+
+	pcie_scan(&pcie_scan_opt);
+
+	if (pcie.bdf == PCIE_BDF_NONE) {
+		LOG_ERR("Failed to find PCIe device");
+		return -ENODEV;
+	}
+
+	LOG_INF("PCIe: ID 0x%08X, BDF 0x%X", pcie.id, pcie.bdf);
+
+	struct pcie_bar mbar_regs, mbar_msi_x, mbar_shmem;
+
+	if (!pcie_get_mbar(pcie.bdf, IVSHMEM_PCIE_REG_BAR_IDX, &mbar_regs)) {
+		LOG_ERR("PCIe register bar not found");
+		return -EINVAL;
+	}
+
+	pcie_set_cmd(pcie.bdf, PCIE_CONF_CMDSTAT_MEM |
+		PCIE_CONF_CMDSTAT_MASTER, true);
+
+	bool msi_x_bar_present = pcie_get_mbar(pcie.bdf, IVSHMEM_PCIE_MSI_X_BAR_IDX, &mbar_msi_x);
+	bool shmem_bar_present = pcie_get_mbar(pcie.bdf, IVSHMEM_PCIE_SHMEM_BAR_IDX, &mbar_shmem);
+
+	LOG_INF("MSI-X bar present: %s", msi_x_bar_present ? "yes" : "no");
+	LOG_INF("SHMEM bar present: %s", shmem_bar_present ? "yes" : "no");
+
+	z_phys_map((uint8_t **)&dev_data->ivshmem_regs,
+		mbar_regs.phys_addr, mbar_regs.size,
+		K_MEM_CACHE_WB | K_MEM_PERM_RW);
+
+	LOG_INF("Registers at 0x%lX (mapped to 0x%lX), size 0x%04zX",
+			mbar_regs.phys_addr, (uintptr_t)dev_data->ivshmem_regs, mbar_regs.size);
+
+	LOG_INF("ivshmem_regs: id %u, max_peers %u",
+		dev_data->ivshmem_regs->id, dev_data->ivshmem_regs->max_peers);
+	if (dev_data->ivshmem_regs->id > 1) {
+		LOG_ERR("Invalid ivshmem ID %u", dev_data->ivshmem_regs->id);
+		return -EINVAL;
+	}
+	if (dev_data->ivshmem_regs->max_peers != 2) {
+		LOG_ERR("Invalid ivshmem max peers %u", dev_data->ivshmem_regs->max_peers);
+		return -EINVAL;
+	}
+	dev_data->peer_id = (dev_data->ivshmem_regs->id == 0) ? 1 : 0;
+
+	uint32_t vendor_cap = pcie_get_cap(pcie.bdf, PCI_CAP_ID_VNDR);
+
+	uintptr_t shmem_phys_addr;
+	uint32_t cap_pos;
+
+	if (shmem_bar_present) {
+		shmem_phys_addr = mbar_shmem.phys_addr;
+	} else {
+		cap_pos = vendor_cap + IVSHMEM_CFG_ADDRESS / 4;
+		shmem_phys_addr = pcie_conf_read_u64(pcie.bdf, cap_pos);
```

Comment:
what specs are you following here? at least it is not in original ivshmem specs.

Reply:
Jailhouse "ivshmem-v2":

https://github.com/siemens/jailhouse/blob/630001202caec85370fb4f956e581f51109e490c/Documentation/ivshmem-v2-specification.md